### DISCUSSION TITLE
BigInt gets rounded off in returned server response

### DISCUSSION THREAD

**Discussion by NexWasTaken** [MichalLytek/typegraphql-prisma#227] [Q&A]
I'm storing IDs in my mysql db as `BigInt`. Whenever I try to query my db, the returned data has the ids rounded off (`718775388976709672` turns to `718775388976709600`). I checked the actual table itself and it has the correct ID (`718775388976709672`) but whenever I query it using gql, it gets rounded off.

How can I fix this issue? And is this related to typegraphql-prisma or Apollo?

**Comment by MichalLytek** [comment]
Does it also happen with using only Prisma Client? Maybe your Prisma schema is incorrect.
`typegraphql-prisma` is using bigint scalar for returning long ints as string.

**Reply by NexWasTaken** [reply]
It does not happen when I query directly with the Prisma Client.
*Using PrismaClient:*
![image](https://user-images.githubusercontent.com/40592339/146651534-e53498fe-3876-4266-a461-72e15ab7f8ee.png)
![image](https://user-images.githubusercontent.com/40592339/146651540-64857b00-23e0-4750-b588-b9a7b15079b0.png)
*Using GraphQL:*
![image](https://user-images.githubusercontent.com/40592339/146651597-c8798777-0673-4170-9148-7c62ae841816.png)

Just a random guess, but perhaps it's an issue with the JSON serialization?

**Reply by MichalLytek** [reply]
You should pass the big integer as a string: `server_id: "718775388976709672"`. JSON is not capable of transporting such big numbers, its default serialization treats numbers using `Number` so it drops the precision of big int.

**Reply by MichalLytek** [reply]
I've checked and when you have `bigint` field in Prisma schema, it properly generates both model and input types with `GraphQLScalars.BigIntResolver`.

**Reply by NexWasTaken** [reply]
How will I pass in the BigInt as a string if the generated type is of BigInt? The issue is not that I have the incorrect generated type, it's that despite of having the type as BigInt it still loses it's precision.

**Reply by MichalLytek** [reply]
JSON is not capable of transferring big integer in a number format.

If you pass the long int as a number, it's get "rounded off". You need to pass it as a string, so that it can then parse that string into `BigInt`.

![image](https://user-images.githubusercontent.com/10618781/146680409-e1c22294-2bb3-4fd9-a1d8-1078998283bd.png)

**Reply by NexWasTaken** [reply]
![image](https://user-images.githubusercontent.com/40592339/146680901-29bef4b2-16ae-47a7-9126-98ec52f1bcab.png)

**Reply by NexWasTaken** [reply]
I'll repeat once again that there's no issue on the client side. I can make valid queries using Number on the client-side and they get parsed correctly as BigInt in the database (see my first ss). 

The issue is specifically in the JSON that gets RETURNED from the server.

**Comment by MichalLytek** [comment]
Ok I can reproduce the issue:
```ts
await prisma.nativeTypeModel.create({
  data: {
    bigInt: BigInt("123456789123456789"),
  },
});
```

![image](https://user-images.githubusercontent.com/10618781/147092751-68f558ab-f162-4082-b6d0-64ce9cf9d86f.png)

```json
{
  "data": {
    "nativeTypeModel": {
      "bigInt": 123456789123456780
    }
  }
}
```

However, it's an issue with the GraphQL Scalar:
https://github.com/Urigo/graphql-scalars/blob/master/src/scalars/BigInt.ts

It should return the BigInt as a string `"123456789123456789"` in JSON, not as a number format `123456789123456780`

So please report the bug on their repo and link to this issue 😉

**Reply by NexWasTaken** [reply]
In the documentation for [GraphQL Scalars](https://www.graphql-scalars.dev/docs/scalars/big-int), it mentions you need to install [`json-bigint-patch`](https://github.com/ardatan/json-bigint-patch) to properly support serializing BigInts

**Reply by MichalLytek** [reply]
@NexWasTaken I've checked that and while it works well for `JSON.parse`, it does not fix the issue with Apollo Server.

Try to reproduce the issue without Prisma and report the bug on their repo. I can't do anything more apart from using other bigint scalar.
